### ISSUE TITLE
strlen primitive

### ISSUE THREAD

**Issue by rkoeninger** [Shen-Language/shen-sources#49]
So this is an issue which is more or less severe depending on your target platform, but in .NET, I found it to be troublesome.

---

### Example in the kernel sources

Some of the functions in the kernel are written to use errors and the `trap-error` expression as a form of flow control, for instance, in `writer.shen`:

```shen
(define pr
  String Sink -> (trap-error
                  (prh String Sink 0)
                  (/. E String)))

(define prh
  String Sink N -> (prh String Sink (write-char-and-inc String Sink N)))

(define write-char-and-inc
  String Sink N -> (do (write-byte (string->n (pos String N)) Sink)
                       (+ N 1)))
```

This code walks down a string, writing each character, until is tries to access a character index past the end of the string, which raises an error in the `pos` function. The error handling construct `trap-error` catches it further up the stack in the `pr` function and then `pr` returns successfully. Notice `prh` recurses with no base case. And it catches *any* error and considers that error an indication of success even if it wasn't just an index-out-of-bounds error.

---

### .NET specific experience

The Visual Studio debugger by default stops execution whenever an exception is raised in your code. Since `pr` raises error internally all the time, this made the debugger unusable.

"Break on first chance exception" can be disabled, but I'd still want to use it for errors that are actually errors. I might also want to set a breakpoint in my `simple-error` definition, which would again be unusable.

Early on in building my F#/.NET port, I had a `Result = Success of Value | Error of string` type to model KL errors. I reverted to just using .NET's built-in exceptions and `try`/`catch` construct as it was simpler and exceptions could pop up or unanticipated reasons. I seem to remember that performance was better with the `Result` type, though, because there wasn't frivolous use of exceptions. Raising an exception has a much greater performance impact than returning a value and has to circumvent optimized code.

---

### Suggested changes

Another way `pr` could be written is to have it print the first character of the string and then recurse on the `tlstr`, returning when it is asked to print `""`. This could be very wasteful, depending on the platforms implementation of a string.

We could also have a `strlen` function that takes a string and returns its length so we know how many characters to print. But with the current primitve set, it would have to implemented as recurse-until-error or tlstr-recurse-until-empty, neither are great approaches.

I think the best solution would be to start requiring an additional primitive: `strlen` with takes a string and returns its length. This would be very easy to implement and would simplify string handling. There may be other places in the code that would benefit.

I know the goal of Shen is to keep the primitive set small, but some compromises made in doing so may not have been worth it. Leaving out a simple primitive like `strlen`, or having `symbol?` be a primitive with a more reasonable implementation, has made porting more difficult than implementing those primitives, in my experience.

**Comment by tizoc** [comment]
There are many cases where a function provided in the Shen kernel is not too great in terms of performance, but thats expected, and the solution in such cases is for the port to override it (this is something the spec itself talks about).

Rather than trying to figure out solutions that may work well in some implementations but not for others I think it is better to just document these cases in the wiki so that porters can know about them and handle them properly.

For this case in particular, even with `strlen` the implementation would be much much worse than a natively implemented one that just writes the whole string to the output port, instead of doing so byte by byte.

Other than that.. an issue I have with `strlen` (that has a bigger weight than the addition of a new primitive) is that it is a good and cheap for implementations that implement strings as a block of contiguous memory, but horrible for the ones that implement them as linked lists (like Haskell does, and which I think may work best in Shen given how it handles string pattern matching). I don't implement it that way in shen-scheme but it is on my list of things to try.

**Comment by rkoeninger** [comment]
The only serious Haskell port is shentong, [which uses Data.Text](https://github.com/mthom/shentong-core/blob/0f96b9af51224a32f191d6b6bac5a1691fec54b0/Types.hs), which I think is considered the normal way to do strings in Haskell these days. I think the Haskellers decided linked lists were too inefficient for most string operations.

And the [definition of Text](https://hackage.haskell.org/package/text-1.2.2.2/docs/src/Data-Text-Internal.html#Text), where it's like String in Java, C# in that it has an array, an offset and length values. And [the tail functions returns a Text with the same array and off+1, len-1](https://hackage.haskell.org/package/text-1.2.2.2/docs/src/Data-Text.html#tail).

---

Anyway, I realize we're trying to strike a balance between keeping the kernel small and simple and being helpful toward port developers (e.g. pre-optimizing KL, `dict`s, `filter`, `fold-left` functions) and provide decent defaults. There's a line to be drawn but I'm not sure where.

Having a function or set of functions in there whose default behavior is going to have to be overwritten by everyone anyway seems odd. It's like, it should have a reasonable default - not raising and catching errors constantly - or should be a primitive.

I feel like this is part of a bigger conversation about the parameters of Shen and I don't want this single issue to grow any further out of scope. We'll keep this thread about `strlen` and say right now we're leaning "no".

**Comment by tizoc** [comment]
Ok, about `strlen` in particular, for now I lean 'no' as you said (as a function in the kernel and primitive, that is), mainly because this particular case doesn't justify it (it is as easy -- and arguably better -- to provide a native implementation of `pr`).

**Offtopic as far as `strlen` goes, but just to answer some things in your comment and share some thoughts:**

Regarding stings as lists, I was talking about standard Haskell (even if "real programs" don't use them), not the port. As things are right now, strings-as-lists is a viable option for a Shen port, one which would make functions that use pattern matching over strings - a common idiom in Shen - faster and less wasteful. Adding a `strlen` primitive kinda favours strings-as-arrays, which I think conflicts with what is in there right now.

I'm not terribly happy with strings in Shen, it is something I discussed more than once in the mailing list years ago, and in the end I just decided that having a library providing a new datatype for strings, along with a bunch of useful (and performant) operations would be good enough, without having to change the kernel.

re: `Data.Text`, thats how I wish Shen strings worked (I proposed something similar long ago, but it didn't catch on), because it really pisses me off that something as basic to the language as `@s` (constructor and in pattern matching) and `tlstr` is so wasteful (same with some vector idioms). Even without fusion like in Haskell it would still be much better than what we have now.

But it is complicated, any change that makes ShenOS diverge too much from ShenProfessional should be considered very carefully.

> Having a function or set of functions in there whose default behaviour is going to have to be overwritten by everyone anyway seems odd. It's like, it should have a reasonable default - not raising and catching errors constantly - or should be a primitive.

Agree, I don't like those `trap-error` as control-flow situations in the kernel, it is ugly and also costly for some ports. Many of such cases disappeared in Shen 20 with the addition of the `*/or` functions, the `pr` is one of those few that remain. I just think that it is easier to just override `pr` than to have a new primitive, and in addition to that the result is better (faster). It is like making `pr` a primitive but it is optional, because there is a default implementation provided (like with dicts).

> There's a line to be drawn but I'm not sure where.

I'm in favour of all such improvements to kernel functionality as long as they don't complicate things too much, and don't have drawbacks (most of the changes I have made to Shen myself are of this kind, changes that make the kernel faster even without ports having to make changes, while also providing hooks for ports to improve performance even more). Same for pre-compilation improvements handled by `make.shen` that just make things faster for everyone without really changing any behaviour.

For anything extra, a library (something still unsolved) is a good place, at least at first, because there is less friction than when introducing something to the kernel (which requires more consideration).

I regret having added some of the functions -- I don't regret stuff like dicts because I think adding them introduced good benefits (also the Kernel already had it's own ad-hoc implementation of dicts internally), same for the`*/or` functions, but stuff like exit, folds, etc could have waited until Shen has some sort of library and placed there instead.
